### ISSUE TITLE
Warning messages to add in the CSV exporter when there's any field dropped

### ISSUE THREAD

**Issue by starrify** [scrapy/scrapy#4002]
## Summary

A warning message shall be generated when there're some fields dropped within `CsvItemExporter` (or also in other exporters?) during exporting.
We may have this kind of warning messages emitted only for the first occurrence to avoid generating too much noise, similar to the warning messages when there's possible data loss ([related code](https://github.com/scrapy/scrapy/blob/d4b8bf18b2c7f563198d4334f5f5622f1006041d/scrapy/core/downloader/handlers/http11.py#L492)).

## Motivation

Due to nature of the CSV format, the `CsvItemExporter` needs to determine the list of fields to write before writing the 1st record ([related code](https://github.com/scrapy/scrapy/blob/d4b8bf18b2c7f563198d4334f5f5622f1006041d/scrapy/exporters.py#L253)), which would in some occasions cause unintentional loss of data.

Unfortunately I've yet observed this kind of mistakenly introduced data loss issue from different developers, where the common cause may be:
- Returning a `dict` as the 1st item that doesn't have all fields included.
- Returning more than one type of items that have different fields.

(both when not having `FEED_EXPORT_FIELDS` properly configured)

## Describe alternatives you've considered

Properly configuring the `FEED_EXPORT_FIELDS` setting entry would surely have the issue resolved. However that's not always guaranteed (a developer may just forget to, etc.).

**Comment by amardeep24** [comment]
@kmike I want to work on this!

**Comment by amardeep24** [comment]
@starrify if `self.fields_to_export` is not set in the `_write_headers_and_set_fields_to_export` method then we should show that warning right? in other words if the  `self.fields_to_export` is populated from the `list(item.keys())` or `list(item.fields.keys())` the warning should indicate the possible loss of data if the `fields_to_export` is not already explicitly set in prior?

**Comment by Gallaecio** [comment]
> @starrify if `self.fields_to_export` is not set in the `_write_headers_and_set_fields_to_export` method then we should show that warning right? in other words if the `self.fields_to_export` is populated from the `list(item.keys())` or `list(item.fields.keys())` the warning should indicate the possible loss of data if the `fields_to_export` is not already explicitly set in prior? @kmike @Gallaecio @elacuesta thoughts?

Correct. But only once an item is found that has more than those fields, and the warning should only be printed once, when the first such item is found.

**Comment by amardeep24** [comment]
@Gallaecio Please confirm my understanding:

1. The` self.fields_to_export` is set with a list of columns and then if **any of the rows have an extra data column the warning is to be displayed once.**

2. When the ` self.fields_to_export` **is not set**, then also we should show the warning once.

**Comment by Gallaecio** [comment]
@amardeep24 I believe that’s correct, yes.

**Comment by kmike** [comment]
I'm not sure about a case where FIELD_EXPORT_FIELDS option is set - if an user asks for a subset of fields explicitly, why should we show a warning? One INFO-level log line can be fine in this case, or we can leave it as-is for now.

**Comment by amardeep24** [comment]
@kmike In that case it just boils down to adding a warning if the `FIELD_EXPORT_FIELDS`  is not set initially, as afterwards it would be set from the `item.keys()` or `item.fields.keys()` :
```
if not self.fields_to_export:
                logger.warning("Possible chance of data loss as FIELD_EXPORT_FIELDS is not set "
                               " -- This message won't be shown in further requests")
                if isinstance(item, dict):
                    # for dicts try using fields of the first item
                    self.fields_to_export = list(item.keys())
                else:
                    # use fields declared in Item
                    self.fields_to_export = list(item.fields.keys())
            row = list(self._build_row(self.fields_to_export))
```

**Comment by Gallaecio** [comment]
But we only want such a warning logged if there is actual data loss, I think.

**Comment by amardeep24** [comment]
@Gallaecio & @kmike In that case we should take care of both **1 & 2** and the below fix should work:

```
if not self.fields_to_export:
    self._is_data_loss_warning_displayed = True
    if isinstance(item, dict):
        # for dicts try using fields of the first item
         self.fields_to_export = list(item.keys())
    else:
         # use fields declared in Item
         self.fields_to_export = list(item.fields.keys())
 elif not self._fail_on_dataloss_warned:
         if isinstance(item, dict):
             if len(item.keys()) > self.fields_to_export:
             self._is_data_loss_warning_displayed = True
        else:
             if len(item.fields.keys()) > self.fields_to_export:
             self._is_data_loss_warning_displayed = True
if self._is_data_loss_warning_displayed and not self._fail_on_dataloss_warned:
    logger.warning("Possible chance of data loss as FIELD_EXPORT_FIELDS is not set "
                              " -- This message won't be shown in further requests")
    self._fail_on_dataloss_warned = True 
```
Here `_is_data_loss_warning_displayed ` is a new flag to indicate whether we should log the warning and `self._fail_on_dataloss_warned` is another new flag to check if the warning was logged previously.

These two glags are initialized the `__init__` method of `CsvItemExporter` to `False`.

Please review so that I can raise a new PR.

**Comment by Gallaecio** [comment]
> Please review so that I can raise a new PR.

Please create a pull request so that we can review :slightly_smiling_face:

**Comment by amardeep24** [comment]
@Gallaecio done!

**Comment by jtboing** [comment]
Has this issue been resolved? If so, then maybe this issue can be closed.

**Comment by moduluz** [comment]
Hi all, I see that this issue (PR #4053) has been 'stuck' for a couple years now. This is what I would love to do as my very first contribution to OS and get it to the finish line!

I intend to incorporate the main logic of that PR though I will:

As pointed out by @ejulio, avoid the use of != to return false positive warnings if the item only has fewer fields, use <= instead.

Remove the redundant if/else block and replace the warnings that @Gallaecio and @elacuesta requested to make the user clear what they should do, which should be to set FEED_EXPORT_FIELDS.

I'm going to fork the repo and create a new PR for these fixes in the near future.
